lib/server/routes/buckets.js

```diff
@@ -1396,6 +1398,129 @@ BucketsRouter.prototype.getStorageUsage = function (req, res) {
   });
 };
 
+BucketsRouter.prototype.finishUpload = function (req, res, next) {
+  const { Bucket, BucketEntry, Upload, Shard, Mirror } = this.storage.models;
+
+  const idBucket = req.params.id;
+
+  const { index, shards } = req.body;
+
+  if (!isHexString(index) || index.length !== 64) {
+    return next(new errors.BadRequestError('Invalid index'));
+  }
+
+  if (!shards) {
+    return next(new errors.BadRequestError('Missing shards'));
+  }
+
+  if (!Array.isArray(shards)) {
+    return next(new errors.BadRequestError('Shards is not an array'));
+  }
+
+  shards.forEach(({ uuid, hash }) => {
+    if (!validateUUID(uuid)) {
+      return next(new errors.BadRequestError('Invalid UUID'));
+    }
+    if (!hash) {
+      return next(new errors.BadRequestError('Missing hash'));
+    }
+  });
+
+  const createdShardIDs = [];
+
+  Bucket.findOne({
+    _id: idBucket,
+    user: req.user._id
+  }, (err, bucket) => {
+    if (err) {
+      return next(new errors.InternalError(err.message));
+    }
+
+    if (!bucket) {
+      return next(new errors.NotFoundError('Bucket not found'));
+    }
+
+    async.eachLimit(
+      shards,
+      1,
+      (shard, cb) => {
+        const { uuid, hash } = shard;
+
+        Upload.findOne({
+          uuid
+        }, (err, upload) => {
+          if (err) {
+            return cb(err);
+          }
+
+          if (!upload) {
+            return cb(new errors.NotFoundError('Upload not found'));
+          }
+
+          const uploadObject = upload.toObject();
+          const { _id, contracts, ...uploadWithoutIdAndContracts } = uploadObject;
+
+          const shardToCreate = {
+            ...uploadWithoutIdAndContracts,
+            hash,
+            contracts: contracts.map(contract => ({
+              ...contract,
+              data_hash: hash
+            })),
+          };
+
+          Shard.create(shardToCreate, (err, savedShard) => {
+            if (err) {
+              return cb(err);
+            }
+
+            createdShardIDs.push(savedShard._id);
+
+            async.eachLimit(
+              contracts,
+              1,
+              (contract, cb) => {
+                const mirrorToCreate = new Mirror({
```

Comment:
Unsure here if its ok to use `new ` or better to create the method inside storage models

Reply:
The chosen option with `new` is fine